### PR TITLE
[opengl] random use different seed in each launch

### PR DESCRIPTION
**Pull request by archibate**
So that `examples/sdf_renderer.py` works.

Related issue = #492 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by yuanming-hu** [comment]
Thanks! I'll test it on my end. Meanwhile, let's finish https://github.com/taichi-dev/taichi/pull/666 first so that we get rid of the redundant commits.

File: taichi/backends/opengl/opengl_api.cpp [hunk 1]
```diff
@@ -236,7 +244,8 @@ GLSSBO *root_ssbo;
 void create_glsl_root_buffer(size_t size) {
   // if (root_ssbo) return;
   initialize_opengl();
-  root_ssbo = new GLSSBO;  // TODO(archibate): mem leaking
+  root_ssbo = new GLSSBO;  // TODO(archibate): mem leaking, use std::optional instead
```
Comment (archibate) [inline taichi/backends/opengl/opengl_api.cpp:247]: A very little change (< 10 CoL) can fix this mem leakage, should I put in this PR?

Comment (yuanming-hu) [reply taichi/backends/opengl/opengl_api.cpp:247]: If it's just 10 LoC then let's fix it here.

**Comment by archibate** [comment]
We may also want to add a test case for this, to check if `ti.random()` return same when called again. e.g.
```py
@ti.kernel
def func(i: ti.i32):
  x[i] = ti.random()
for i in range(5):
  func(i)
  assert x[i] != x[i - 1]
```

**Comment by yuanming-hu** [comment]
> We may also want to add a test case for this, to check if `ti.random()` return same when called again. e.g.
> 
> ```python
> @ti.kernel
> def func(i: ti.i32):
>   x[i] = ti.random()
> for i in range(5):
>   func(i)
>   assert x[i] != x[i - 1]
> ```

This might be a brittle test since there's a tiny probability that `x[i] == x[i - 1]`. It will be better to, say repeat 5 times and assert there are at least 4 different results.

**Comment by yuanming-hu** [comment]
Screenshot from 2020-04-02 12-00-36

I confirm this works on my end. It reaches 90 samples per pixel per second, which is 1.8x faster than the CUDA backend!

**Comment by archibate** [comment]
That's amazing! Couldn't believe we are now faster than CUDA! Does this mean that CUDA still has lot of space of optimizations to be done?

**Comment by yuanming-hu** [comment]
I guess it's because CUDA is not yet fully optimized (and the OpenGL backend is well done). I haven't got a chance to work on it.

**Comment by archibate** [comment]
Seems #603 is just 杞人忧天ing... the performance of current OpenGL backend is really utilized now.
NEXT: to support sparse data structures on GL, so that the thought-to-be-weak GLSL can really 平起平坐 with CUDA :)

**Comment by archibate** [comment]
#603 says:

> Most time costed on sync (data transfer from&to GPU) thanks to GL's great API design, we may want to figure out how to use it more efficiently.

In fact I found sync is only costly on mesa one, not NV, maybe their GPUs have really faster communication with CPU?

**Comment by yuanming-hu** [comment]
> #603 says:
> 
> > Most time costed on sync (data transfer from&to GPU) thanks to GL's great API design, we may want to figure out how to use it more efficiently.
> 
> In fact I found sync is only costly on mesa one, not NV, maybe their GPUs have really faster communication with CPU?

I guess it's because the NVIDIA driver is better implemented than mesa for NVIDIA GPUs.

**Comment by archibate** [comment]
Could you also test `examples/nbody_oscillator.py` on NV-GL & CUDA? I guess CUDA is better when dealing with memory intensive kernels (doesn't it?)

**Comment by yuanming-hu** [comment]
I have to change `N` to `80000` to make a difference. Then OpenGL reaches 15 FPS and CUDA 5 FPS. OpenGL is 3x faster :-) It seems me that there's a performance bug in the CUDA backend, which I should fix as soon as possible...

**Comment by yuanming-hu** [comment]
I updated https://github.com/taichi-dev/taichi/pull/692/files#diff-733c00d9e85a14e095b788b8b985e75c

Btw do you want to fix the memory leakage here? If not I'll go merge this PR.

**Comment by archibate** [comment]
> It seems me that there's a performance bug in the CUDA backend, which I should fix as soon as possible...

Sounds a serious issue, good luck!

> Btw do you want to fix the memory leakage here? If not I'll go merge this PR.

No, let's just keep this PR small and trackable :)

**Comment by yuanming-hu** [comment]
Cool, thanks. Merging this in now...

File: taichi/backends/opengl/opengl_api.cpp (hunk 1 above)
Comment (archibate) [reply taichi/backends/opengl/opengl_api.cpp:247]: Ok.

**Pull request merged**